contracts/tournament/Tournament.cairo

```diff
@@ -327,4 +430,119 @@ func _only_tournament_closed{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, r
         assert is_tournament_open = FALSE
     end
     return ()
-end
\ No newline at end of file
+end
+
+# Recursively plays all battles until there is only one winner
+func _rec_start{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}():
+    let (ship_count) = playing_ships_count_.read()
+    assert_not_zero(ship_count)
+    
+    _rec_play_current_round_battles(0)
+    _update_playing_ships()
+
+    let (ship_count) = playing_ships_count_.read()
+    if ship_count == 1:
+        # This is the end! We have a single winner!
+        return ()
+    end
+
+    _rec_start()
+    return ()
+end
+
+# Recursively plays all battles
+func _rec_play_current_round_battles{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(battle_ship_index : felt):
+    
+    let (end_reached : felt, next_battle_ship_index : felt) = _play_next_battle(battle_ship_index)
+    if end_reached == TRUE:
+        return ()
+    end
+    
+    _rec_play_current_round_battles(next_battle_ship_index)
+    return ()
+end
+
+# Play the next battle entirely
+func _play_next_battle{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(ship_index : felt) -> (end_reached : felt, next_battle_ship_index : felt):
```

Comment:
I had hard times understanding the logic of this function, I guess a more precise name would have helped a lot, something like `tournament_end_reached` would be a better fit, or anything else that tells the same story, `last_battle_ended`, `all_battles_done` etc

Reply:
I refactored this function that was doing multiple things. Now it should be clearer.